File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from math import sqrt
from itertools import permutations
from collections import defaultdict

isPrime = []

def calcPrimes():
    global isPrime
    
    maxNum = 9999999
    isPrime = [True] * (maxNum + 1)
    isPrime[0] = isPrime[1] = False
    
    for i in range(2, int(sqrt(maxNum)) + 1):
        if not isPrime[i]: continue
        for j in range(i * i, maxNum + 1, i):
            isPrime[j] = False

def solution(numbers):
    calcPrimes()
    
    visited = defaultdict(bool)
    answer = 0
    
    for i in range(1, len(numbers) + 1):
        for perm in permutations(numbers, i):
            number = int(''.join(perm))
            if visited[number] or not isPrime[number] : continue
            visited[number] = True
            answer += 1
    
    return answer

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

vector<bool> isPrime;
unordered_map<int, bool> mp;

void eratos(int maxNum) {
    isPrime.resize(2);
    isPrime.resize(maxNum + 1, true);
    for (int i = 2; i * i <= maxNum; i++) {
        if (!isPrime[i]) continue;
        for (int j = i * i; j <= maxNum; j += i) {
            isPrime[j] = false;
        }
    }
}

void permute(int visited, string cur, const int maxDepth, const string &base) {
    if (cur.size() == maxDepth) {
        int num = stoi(cur);
        if (isPrime[num]) mp[num] = true;
        return;
    }
    
    for (int i = 0; i < base.size(); i++) {
        if (visited & (1 << i)) continue;
        permute(visited | (1 << i), cur + base[i], maxDepth, base);
    }
}

int solution(string numbers) {
    eratos(9999999);
    for (int i = 1; i <= numbers.size(); i++) {
        permute(0, {}, i, numbers);
    }
    int answer = mp.size();
    return answer;
}